pkg/controller/master/schedulevmbackup/common.go

```diff
@@ -0,0 +1,480 @@
+package schedulevmbackup
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	"go.uber.org/multierr"
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/harvester/pkg/controller/master/backup"
+	"github.com/harvester/harvester/pkg/ref"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	scheduleVMBackupKindName = "ScheduleVMBackup"
+	timeFormat               = "20060102.1504"
+
+	reachMaxFailure = "Reach Max Failure"
+
+	proactiveSuspend = "Proactive Scheulde Suspend"
+)
+
+const (
+	updateInterval    = 5 * time.Second
+	reconcileInterval = 5 * time.Second
+
+	svmbackupPrefix = "svmb"
+
+	cronJobNamespace    = "harvester-system"
+	cronJobBackoffLimit = 3
+	cronJobCmd          = "sleep"
+	cronJobArg          = "10"
+)
+
+var scheduleVMBackupKind = harvesterv1.SchemeGroupVersion.WithKind(scheduleVMBackupKindName)
+
+func cronJobName(svmbackup *harvesterv1.ScheduleVMBackup) string {
+	return fmt.Sprintf("%s-%s", svmbackupPrefix, svmbackup.UID)
+}
+
+func getCronJob(h *svmbackupHandler, svmbackup *harvesterv1.ScheduleVMBackup) (*batchv1.CronJob, error) {
+	cronJob, err := h.cronJobCache.Get(cronJobNamespace, cronJobName(svmbackup))
+	if err != nil {
+		return nil, err
+	}
+
+	return cronJob, nil
+}
+
+func vmBackupName(svmbackup *harvesterv1.ScheduleVMBackup, timestamp string) string {
+	return fmt.Sprintf("%s-%s-%s", svmbackupPrefix, svmbackup.UID, timestamp)
+}
+
+func getVMBackup(h *svmbackupHandler, svmbackup *harvesterv1.ScheduleVMBackup, timestamp string) (*harvesterv1.VirtualMachineBackup, error) {
+	vmBackup, err := h.vmBackupCache.Get(svmbackup.Namespace, vmBackupName(svmbackup, timestamp))
+	if err != nil {
+		return nil, err
+	}
+
+	return vmBackup, nil
+}
+
+func resolveSVMBackupRef(h *svmbackupHandler, obj metav1.Object) *harvesterv1.ScheduleVMBackup {
+	var annotations = obj.GetAnnotations()
+	if annotations == nil || annotations[util.AnnotationSVMBackupID] == "" {
+		return nil
+	}
+
+	namespace, name := ref.Parse(annotations[util.AnnotationSVMBackupID])
```

Comment:
We may need to add webhook for this annotation to prevent users from updating it.

Reply:
If the old `svmbackup`  CR is already in a resume state, the Harvester controller doesn't need to check backup target connectivity.

If the new `svmbackup` CR asks to suspend the schedule, the Harvester controller doesn't need to check backup target connectivity, either.

Reply:
Updated per suggestion, thanks.